src/Microsoft.Azure.EventHubs/EventHubsDiagnosticSource.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.EventHubs
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Diagnostic Source util for "Microsoft.Azure.EventHubs" source.
+    /// </summary>
+    internal static class EventHubsDiagnosticSource
+    {
+        public const string DiagnosticSourceName = "Microsoft.Azure.EventHubs";
+
+        public const string SendActivityName = DiagnosticSourceName + ".Send";
+        public const string SendActivityStartName = SendActivityName + ".Start";
+        public const string SendActivityExceptionName = SendActivityName + ".Exception";
+
+        public const string ReceiveActivityName = DiagnosticSourceName + ".Receive";
+        public const string ReceiveActivityStartName = ReceiveActivityName + ".Start";
+        public const string ReceiveActivityExceptionName = ReceiveActivityName + ".Exception";
+
+        internal static readonly DiagnosticListener DiagnosticListener = new DiagnosticListener(DiagnosticSourceName);
+
+        internal static bool IsEnabled => DiagnosticListener.IsEnabled();
+
+        internal static Activity StartSendActivity(string clientId, EventHubsConnectionStringBuilder csb, string partitionKey, IEnumerable<EventData> eventDatas, int count)
+        {
+            // skip if diagnostic source not enabled
+            if (!IsEnabled)
+            {
+                return null;
+            }
+
+            // skip if no listeners for this "Send" activity 
+            if (!DiagnosticListener.IsEnabled(SendActivityName,
+                new
+                {
+                    Endpoint = csb.Endpoint,
+                    EntityPath = csb.EntityPath,
+                    PartitionKey = partitionKey
+                }))
+            {
+                return null;
+            }
+
+            Activity activity = new Activity(SendActivityName);
+
+            // extract activity tags from input
+            activity.AddTag("component", "Microsoft.Azure.EventHubs");
+            activity.AddTag("span.kind", "producer");
+            activity.AddTag("operation.name", $"Send");
+            activity.AddTag("operation.data", $"{csb.EntityPath}/{partitionKey}");
+            activity.AddTag("peer.service", "Azure Event Hub");
+            activity.AddTag("peer.hostname", csb.Endpoint.OriginalString);
+            activity.AddTag("eh.event_hub_name", csb.EntityPath);
+            activity.AddTag("eh.partition_key", partitionKey);
+            activity.AddTag("eh.event_count", count.ToString());
+            activity.AddTag("eh.client_id", clientId);
```

Comment:
what is this client ID for? Can/should we replace it with the activity ID so event hub server will correlate telemetry with the Activity ID?

Reply:
If not part of this PR - please create an issue to discuss and track it

Reply:
Based on the code, ```ClientId``` is fixed for given client instance and would be the same for numerous requests (thus Activities). It looks to be primarily used for tracing (multiple EventSource events) but also metering (e.g. retry count)